src/lib/navigation/navigation-mode-factory.provider.ts

```diff
@@ -0,0 +1,78 @@
+import {FreeNavigationMode} from './free-navigation-mode';
+import {NavigationMode} from './navigation-mode.interface';
+import {SemiStrictNavigationMode} from './semi-strict-navigation-mode';
+import {StrictNavigationMode} from './strict-navigation-mode';
+import {WizardComponent} from '../components/wizard.component';
+import {NavigationModeInput} from './navigation-mode-input.interface';
+
+/**
+ * A factory used to create [[NavigationMode]] instances
+ */
+export class BaseNavigationModeFactory {
+
+  /**
+   * @inheritDoc
+   */
+  create(wizard: WizardComponent, navigationModeInput: NavigationModeInput): NavigationMode {
+    let navigationModeName: string;
+    if (typeof navigationModeInput === 'function') {
+      // input is a function
+      const navigationMode = navigationModeInput(wizard);
+      if (typeof navigationMode === 'string') {
+        // function returned a name
+        navigationModeName = navigationMode;
+      } else {
+        // function returned a `NavigationMode` instance
+        return navigationMode;
+      }
+    } else {
+      // input is a name
+      navigationModeName = navigationModeInput;
+    }
+    // create NavigationMode by name
+    return this.createByName(wizard, navigationModeName);
+  }
+
+  /**
+   * Create a [[NavigationMode]] for the given wizard instance by a navigation mode name
+   *
+   * @param wizard The wizard componenent where the created [[NavigationMode]] will be used
+   * @param navigationModeInput The name of a built-in navigation mode or a custom navigation mode
+   * @returns The created [[NavigationMode]]
+   */
+  createByName(wizard: WizardComponent, navigationMode: string): NavigationMode {
```

Comment:
Add `private`?

Reply:
I guess, `protected` will be better so that this method can be overridden in a custom navigation mode factory.

Reply:
72bae52